### PR TITLE
[fix][broker] Fix unloadNamespaceBundlesGracefully can be stuck with extensible load manager 

### PR DESCRIPTION
**Pull request by BewareMyPower**
### Motivation

I observed an issue that broker was stuck at close for a long time. It's stuck at `BrokerService#unloadNamespaceBundlesGracefully`, which calls `disableBroker` once and `unloadNamespaceBundleAsync` for all owned namespace bundles **synchronously**. Most issues happen when the broker is the last broker.

#### Issue 1: Free events won't be sent in `overrideOwnership`

In `overrideOwnership`, if no broker is available, a `Free` event will be created.

https://github.com/apache/pulsar/blob/4ce0c752cc4b2d6dccb818ab0ffa854e82e42b85/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java#L1363

However, since the `dstBroker` and `sourceBroker` fields are null in the `Free` event, an exception will be thrown so that the `Free` event won't be created and sent.

https://github.com/apache/pulsar/blob/4ce0c752cc4b2d6dccb818ab0ffa854e82e42b85/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateData.java#L37

#### Issue 2: Free events could be skipped due to the same version id

The `Free` event is created in `overrideOwnership` based on a previous event on the same bundle from the table view. However, there might be inflight events that are not in the table view yet. In `ServiceUnitStateDataConflictResolver#shouldKeepLeft`, if the version id is the same, the `Free` event will be skipped. Then, if the last event is the `Owned` event whose target broker is the current broker in close, `waitForCleanups` will wait until the timeout exceeds.

https://github.com/apache/pulsar/blob/4ce0c752cc4b2d6dccb818ab0ffa854e82e42b85/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java#L1394-L1396

#### Issue 3: `__change_events` topic prevents `waitForCleanups` from finishing

The `__change_events` topic's reader, which is managed by the system topic based topic policies service, will try acquire the ownership. So that in `waitForCleanups`, there will always be a `Owned` event for this topic's bundle. If the target broker is the broker itself, `waitForCleanups` will never have a chance to exit until the timeout exceeds.

#### Issue 4: unloadNamespaceBundleAsync will be stuck at getOwnershipAsync if there is no available broker

Broker unregisters itself in `disableBroker`, if it's the last broker, then no brokers will be available after that. However, `unloadNamespaceBundleAsync` needs to publish a `Unloaded` event in `unloadAsync`.

https://github.com/apache/pulsar/blob/4ce0c752cc4b2d6dccb818ab0ffa854e82e42b85/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java#L700-L701

### Modifications

- issue 1: change the constructor of `ServiceUnitStateData` to all null `dstBroker` and `sourceBroker` for `Free` events.
- issue 2: call `TableView#refreshAsync` to refresh the entry set in `ServiceUnitStateTableViewImpl#flush` and call it  before `overrideOwnership`.
- issue 3: close the topic policies service at the beginning of `PulsarService#closeAsync`.
- issue 4: close the extensible load manager if there is no available broker after `unregister`

In addition, in `disableBroker`, cancel the load data report tasks and shutdown the `LoadDataStore` objects to avoid being affected by the producers and readers on these two non-persistent topics.

Since `LoadDataStore#get` is still used in `LeastResourceUsageWithWeight#select`, don't throw an exception and return an empty in `get`. And handle the case that `select` might throw an exception in `ExtensibleLoadManagerImpl#selectAsync`.

To handle the specific case when the broker is the last broker, close the broker in advance if there is no available broker in the metadata store. Then any namespace bundle's unload will also be skipped because the state was restored to INIT.

Add `testLookup` to cover the changes above.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Review by lhotari — APPROVED** [review]
LGTM

**Comment by heesung-sn** [comment]
Thank you for taking a loot at this issue.


For the last broker shutdown issue,

I think we have two options:

1. set all ownerships to Free and shutdown

This can be a little tricky because the clients could trigger lookups and new assignments might happen during shutdown. We can introduce a state to fail assignment during shutdown.

2. skip the gracefulShutdown and just shutdown loadManager.

Since there is no broker available to transfer ownerships, we could just simply shutdown the last broker -- since the load balancer is shutdown, no new assignment will happen during shutdown.

When the fist broker(leader) starts, it will fix any orphan ones. 

Seems like this pr is moving to the option 1.

Regarding the skip message issue,
I think the current skip logic can return lookups too soon, and I dont see a good reason to keep it. For example, when there are concurrent Assign events, it could return deferred lookups too soon by the skip msg logic, before Own event. I think it can just wait for the final Own event.

**Comment by heesung-sn** [comment]
Thank you for finding bugs in the ServiceUnitStateData constructor that does not consider the Free state case and that waitForCleanup should handle Free case properly too.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java [hunk 1]
```diff
@@ -182,6 +182,7 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
     private SplitManager splitManager;
 
     volatile boolean started = false;
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:184]: Can we introduce an enum state?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java [hunk 2]
```diff
@@ -141,6 +141,7 @@ public class ServiceUnitStateChannelImpl implements ServiceUnitStateChannel {
     private volatile long lastOwnEventHandledAt = 0;
     private long lastOwnedServiceUnitCountAt = 0;
     private int totalOwnedServiceUnitCnt = 0;
+    private volatile boolean disablePubOwnedEvent = false;
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:144]: Can we add this case in the ChannelState enum below? I think a better name could be "CleaningOwnerships"

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java [hunk 3]
```diff
@@ -307,7 +309,7 @@ public synchronized void start() throws PulsarServerException {
                             pulsar.getConfiguration().getDefaultNumberOfNamespaceBundles());
 
             tableview = createServiceUnitStateTableView();
-            tableview.start(pulsar, this::handleEvent, this::handleExisting);
+            tableview.start(pulsar, this::handleEvent, this::handleExisting, this::handleSkippedEvent);
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:312]: I am actually not sure if this skip logic is required as this can return deferred lookups too soon.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java [hunk 4]
```diff
@@ -774,7 +774,11 @@ brokerId, getLogEventTag(data), serviceUnit,
         }
     }
 
-    private void handleSkippedEvent(String serviceUnit) {
+    private void handleSkippedEvent(String serviceUnit, ServiceUnitStateData skippedData) {
+        if (skippedData.state() == Free) {
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:778]: This is risky as the skipped message impacts the current states.

The channel design principle is that only non-skipped messages should impact the states in this channel.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java [hunk 5]
```diff
@@ -1381,8 +1400,10 @@ private void overrideOwnership(String serviceUnit, ServiceUnitStateData orphanDa
 
     private void waitForCleanups(String broker, boolean excludeSystemTopics, int maxWaitTimeInMillis) {
         long started = System.currentTimeMillis();
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:1402]: I think we can skip this waiting logic if the broker to cleanup is the last broker. 

I think this waiting logic is optional, and the intention is to give some time to the cluster to converge cleaned ownership view, if ever. 

However, because of the nature of the current ownership changes, there might be a chance that the ownership view might not converge within the timeout -- eventually the leader should fix any orphan or too long inflight states, though).

If there is no broker waiting, we can just skip this waiting.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateData.java [hunk 6]
```diff
@@ -34,7 +34,7 @@ public record ServiceUnitStateData(
 
     public ServiceUnitStateData {
         Objects.requireNonNull(state);
-        if (StringUtils.isBlank(dstBroker) && StringUtils.isBlank(sourceBroker)) {
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateData.java:37]: The intention is that we want to tag a source or destination broker for all events to trace where the event comes from.

For example, the logic tries to set "source broker" when setting Free state, to indicate who was the previous owner. In what cases, both source and dst brokers can be empty here?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java [hunk 7]
```diff
@@ -815,6 +819,15 @@ private void handleOwnEvent(String serviceUnit, ServiceUnitStateData data) {
 
     private void handleAssignEvent(String serviceUnit, ServiceUnitStateData data) {
         if (isTargetBroker(data.dstBroker())) {
+            if (disablePubOwnedEvent) {
+                log.info("Skip assigning self({}) as the owner after cleanOwnerships", serviceUnit);
+                final var getOwnerRequest = getOwnerRequests.remove(serviceUnit);
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:824]: This will make other brokers to wait for the Owned event. I think whatever ongoing state change should follow the protocol without interruption.

I think better place to check this "disablePubOwnedEvent" state in before publishing the state change command.

- publishAssignEventAsync
- publishOverrideEventAsync
- publishUnloadEventAsync
- publishSplitEventAsync

We already have `validateChannelState` func to check the channel state, and I think we can additionally consider "disablePubOwnedEvent" state there.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 2 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:144]: Actually I found there is a `Disabled` state introduced in https://github.com/apache/pulsar/pull/23301. But it's never set other than tests.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 3 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:312]: Let me revert this change.

This skip logic is not required. I restored this logic just because I want to handle the skipped Free events in the same method. But from your explanation [here](https://github.com/apache/pulsar/pull/23349#discussion_r1775957883), handling the skipped Free events might be dangerous so I'm going to think of another solution

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 5 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:1402]: >  if the broker to cleanup is the last broker.

I found sometimes the cleanup did not fail for the last broker but for the 2nd broker in `testCloseAfterLoadingBundles`.

```
2024-09-26T15:35:56,001 - INFO  - [main:BrokerService] - Disable broker in load manager completed in 0.081 seconds
2024-09-26T15:36:31,287 - INFO  - [main:BrokerService] - Disable broker in load manager completed in 35.099 seconds
2024-09-26T15:36:31,644 - INFO  - [main:BrokerService] - Disable broker in load manager completed in 0.214 seconds
2024-09-26T15:36:31,783 - INFO  - [main:ExtensibleLoadManagerCloseTest] - Brokers close time: [275, 35242, 354]
```

Let me look it deeper

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateData.java (hunk 6 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateData.java:37]: https://github.com/apache/pulsar/blob/b1c5d96ab480ba1b8ad8cbe2077cbe4c467dfc61/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java#L1364

If the state is not `Owned`, there is still a case that `orphanData.sourceBroker()` is null. Here is the log I debugged

```
ServiceUnitStateData[state=Assigning, dstBroker=localhost:50930, sourceBroker=null,
```

**Comment by BewareMyPower** [comment]
I reverted the changes to `Free` events and disabled all events except the override events (whose `force` field is true) during `cleanOwnerships()`, PTAL again @heesung-sn

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 2 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:144]: I used this state in the latest code now

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 5 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:1393]: The cause when the 2nd broker failed to close is the issue 5 I mentioned in the latest PR description.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 7 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java]: I used the `Disabled` state instead and disabled all events except the override events (whose force field is true)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java [hunk 8]
```diff
@@ -791,14 +782,36 @@ public void close() throws PulsarServerException {
...
+        }
+        try {
+            brokerLoadDataStore.shutdown();
+        } catch (IOException e) {
+            log.warn("Failed to shutdown brokerLoadDataStore", e);
+        }
+        try {
+            topBundlesLoadDataStore.shutdown();
+        } catch (IOException e) {
+            log.warn("Failed to shutdown brokerLoadDataStore", e);
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:811]: topBundlesLoadDataStore

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java [hunk 9]
```diff
@@ -982,17 +1017,28 @@ protected void monitor() {
...
+        }
         // Close the internal topics (if owned any) after giving up the possible leader role,
         // so that the subsequent lookups could hit the next leader.
         closeInternalTopics();
     }
 
     private void closeInternalTopics() {
         List<CompletableFuture<Void>> futures = new ArrayList<>();
-        for (String name : INTERNAL_TOPICS) {
+        for (String name : Set.of(BROKER_LOAD_DATA_STORE_TOPIC, TOP_BUNDLES_LOAD_DATA_STORE_TOPIC)) {
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:1041]: Could you explain why this change is required?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java [hunk 10]
```diff
@@ -1391,7 +1402,9 @@ private void waitForCleanups(String broker, boolean excludeSystemTopics, int max
                     continue;
                 }
 
-                if (data.state() == Owned && broker.equals(data.dstBroker())) {
+                if (data.state() == Free) {
+                    futures.put(serviceUnit, handleFreeEvent(serviceUnit, data));
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:1406]: why do we need to additionally handleFreeEvent here?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java [hunk 11]
```diff
@@ -679,11 +678,15 @@ private void handleEvent(String serviceUnit, ServiceUnitStateData data) {
                     brokerId, serviceUnit, data, totalHandledRequests);
         }
 
-        if (channelState == Disabled) {
+        ServiceUnitState state = state(data);
+        if (channelState == Disabled && (data == null || !data.force())) {
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:682]: why do we need `(data == null || !data.force()` logic here?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/store/TableViewLoadDataStoreImpl.java [hunk 12]
```diff
@@ -92,7 +92,11 @@ public synchronized CompletableFuture<Void> removeAsync(String key) {
     public synchronized Optional<T> get(String key) {
         String msg = validateTableView();
         if (StringUtils.isNotBlank(msg)) {
-            throw new IllegalStateException(msg);
+            if (msg.equals(SHUTDOWN_ERR_MSG)) {
+                return Optional.empty();
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/store/TableViewLoadDataStoreImpl.java:96]: can you explain why we need to return empty here?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java (hunk 9 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:1041]: See the issue 5 that I added in the PR description.

If the `loadbalancer-service-unit-state` topic's owner happens to be the current broker, the producers could be disconnected. Then in `unloadNamespaceBundleAsync` it would fail to send unload related events.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 10 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:1406]: It might not be needed here now. Let me double confirm it

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java (hunk 9 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java]: It's not required. I reverted this change first.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java (hunk 10 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:1411]: Reverted

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/store/TableViewLoadDataStoreImpl.java (hunk 12 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/store/TableViewLoadDataStoreImpl.java:96]: Without this change, `testCloseAfterLoadingBundles` would always fail.

```
2024-09-26T22:29:35,236 - INFO  - [main:ServiceUnitStateChannelImpl] - Started ownership cleanup for the inactive broker:localhost:58804
2024-09-26T22:29:35,244 - ERROR - [main:ServiceUnitStateChannelImpl] - Failed to override inactiveBroker:localhost:58804 ownership serviceUnit:public/default/0x50000000_0x60000000 orphanData:ServiceUnitStateData[state=Owned, dstBroker=localhost:58804, sourceBroker=localhost:58783, splitServiceUnitToDestBroker=null, force=true, timestamp=1727360975026, versionId=3]. totalCleanupErrorCnt:1
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: This load store tableview has been shutdown
...
```

**Review by heesung-sn — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java [hunk 13]
```diff
@@ -364,8 +364,8 @@ private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Void> f
                                   }
                                }).exceptionally(ex -> {
                                    if (ex.getCause() instanceof PulsarClientException.AlreadyClosedException) {
-                                       log.error("Reader {} was closed while reading existing messages.",
-                                               reader.getTopic(), ex);
+                                       log.info("Reader {} was closed while reading existing messages.",
```
Comment (Demogorgon314) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:367]: Why change the log level from error to info?

**Review by Demogorgon314 — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java (hunk 13 above)
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:367]: To optimize the error logs because the `AlreadyClosedException` is an expected exception but the full stack will be printed each time

**Comment by BewareMyPower** [comment]
It seems there are some failed tests in `ExtensibleLoadManagerImplTest`, please don't merge and let me take a look

**Comment by BewareMyPower** [comment]
image

`ExtensibleLoadManagerImplTest` passed locally now. Let's wait if it would fail.
